### ISSUE TITLE
Restructure message topics

### ISSUE THREAD

**Issue by poelstra** [FirstLegoLeague/mq-protocols#9]
Currently, most (all?) topics used by the FLL apps are using a colon as separator between a 'subsystem' of an app and a 'command/event', e.g. `clock:arm`.
Also, nodes are currently used for specifying an 'app' (e.g. `overlay`).

However, MHub's topic system (and especially the pattern matching for subscribing and e.g. permission management) is based on path-like topics, e.g. `/some/topic`, i.e. separated with slashes.
MHub nodes should also mostly be seen as 'virtual hosts', i.e. most components in a system would normally connect to the `default` node.

The proposal is to change the naming scheme of nodes and topics as used by all current FLL apps to be more consistent with MHub's underlying concepts.

This basically means changing the node name of each app to `default`, and changing the topic 'schema' to something like `/app/subsystem/event`, e.g. `/overlay/clock/arm`, `/scoring/ranking/published/qualifying`.

Advantages:
- better pattern matching possibilities (for subscribe, but also for specifying more detailed publish and subscribe permissions)
- easier to understand that topics can be a 'hierarchy' of events (like the ranking message, above)
- the `default` node is already always available in the default MHub configuration, so for a simple setup it works out-of-the-box
- easier to support 'multi-room' events (by using nodes for the 'rooms', e.g. pit area and practice area, which will have e.g. twitter streams and general announcements in common, but not scoring)
- better compatibility with future features in MHub

Disadvantages:
- need to change the nodes/topics in existing apps (displaySystem, fllscoring, twitterAdmin, more?)
- may have to inform external parties of this change (I think this is only jpschewe's app right now)
- need to update mq-protocols documentation
- using old with new apps will break if we don't have backwards compatible messages

I suppose the work is manageable, and we could even have a simple MHub Relay transform to also forward specific new style messages to their old naming or vice-versa, if necessary.

**Comment by poelstra** [comment]
BTW, reason for bringing it up now, is that I'm going to add a number of events to the fllscoring app, and would like to use the 'correct' naming for them.

**Comment by kmeesters** [comment]
I think this would be a good change, it would make it easier to expand the system in the future. A quick (mental) assessment says we should just carefully examine the existing code to replace the colon with the slash everywhere. In any case, now would be a good time to do it, after the initial release and before we start adding more protocols and potential third party modules.

@idanstark42 : I think you have worked with the MHub messages in your code? Would be able to asses / provide your input on this change?

**Comment by idanstark42** [comment]
That sounds like a very positive change, and if all we talk about is changing the `topic` of the mhub messages to the new format, then it's relatively easy to implement. Notice that if we have a users configuration, like in the launcher release, we have to change it there too, and also in all relevant relays

**Comment by poelstra** [comment]
Yes, it should be changing the default node to `default` and 'moving' the original node name to the topic, and changing colon to slash.

Good point about the relays and other custom configs!

I'd propose to make a list here of all nodes+topics that are currently in use, then come up with a list of what it should be. Maybe we'll also fix an inconsistency here or there, too.

idanstark42 <notifications@github.com> schreef op 7 november 2017 22:28:33 CET:
